protocol-units/bridge/shared/src/bridge_service/active_swap.rs

```diff
@@ -0,0 +1,526 @@
+			}
+			ActiveSwapState::Completed => f.debug_tuple("Completed").finish(),
+			ActiveSwapState::Aborted => f.debug_tuple("Aborted").finish(),
+		}
+	}
+}
+
+#[derive(Debug, Clone)]
+pub struct ActiveSwapConfig {
+	pub error_attempts: usize,
+	pub error_delay: Duration,
+	pub contract_call_timeout: Duration,
+}
+impl Default for ActiveSwapConfig {
+	fn default() -> Self {
+		Self {
+			error_attempts: 3,
+			error_delay: Duration::from_secs(5),
+			contract_call_timeout: Duration::from_secs(30),
+		}
+	}
+}
+
+pub struct ActiveSwapMap<BFrom, BTo>
+where
+	BFrom: BlockchainService,
+	BTo: BlockchainService,
+{
+	pub config: ActiveSwapConfig,
+	pub initiator_contract: BFrom::InitiatorContract,
+	pub counterparty_contract: BTo::CounterpartyContract,
+	swaps: HashMap<BridgeTransferId<BFrom::Hash>, ActiveSwap<BFrom, BTo>>,
+	waker: AtomicWaker,
+}
+
+impl<BFrom, BTo> std::fmt::Debug for ActiveSwapMap<BFrom, BTo>
+where
+	BFrom: BlockchainService,
+	BTo: BlockchainService,
+{
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		f.debug_struct("ActiveSwapMap")
+			.field("swaps", &self.swaps)
+			.field("config", &self.config)
+			.finish()
+	}
+}
+
+#[derive(Debug, Error)]
+pub enum ActiveSwapMapError {
+	#[error("Non existing swap")]
+	NonExistingSwap,
+}
+
+impl<BTo, BFrom> ActiveSwapMap<BFrom, BTo>
+where
+	BTo: BlockchainService + 'static,
+	BFrom: BlockchainService + 'static,
+{
+	pub fn build(
+		initiator_contract: BFrom::InitiatorContract,
+		counterparty_contract: BTo::CounterpartyContract,
+		config: ActiveSwapConfig,
+	) -> Self {
+		Self {
+			initiator_contract,
+			counterparty_contract,
+			swaps: HashMap::new(),
+			config,
+			waker: AtomicWaker::new(),
+		}
+	}
+
+	pub fn get(&self, key: &BridgeTransferId<BFrom::Hash>) -> Option<&ActiveSwap<BFrom, BTo>> {
+		self.swaps.get(key)
+	}
+
+	pub fn get_mut(
+		&mut self,
+		key: &BridgeTransferId<BFrom::Hash>,
+	) -> Option<&mut ActiveSwap<BFrom, BTo>> {
+		self.swaps.get_mut(key)
+	}
+
+	pub fn already_executing(&self, key: &BridgeTransferId<BFrom::Hash>) -> bool {
+		self.swaps.contains_key(key)
+	}
+
+	pub fn start_bridge_transfer(
+		&mut self,
+		details: BridgeTransferDetails<BFrom::Address, BFrom::Hash>,
+	) where
+		BTo::Hash: From<BFrom::Hash>,
+	{
+		assert!(self.swaps.get(&details.bridge_transfer_id).is_none());
+
+		let counterparty_contract = self.counterparty_contract.clone();
+		let bridge_transfer_id = details.bridge_transfer_id.clone();
+
+		tracing::trace!("Starting active swap for bridge transfer {:?}", bridge_transfer_id);
+
+		self.swaps.insert(
+			bridge_transfer_id,
+			ActiveSwap {
+				details: details.clone(),
+				state: ActiveSwapState::LockingTokens(
+					call_lock_bridge_transfer_assets::<BFrom, BTo>(counterparty_contract, details)
+						.boxed()
+						.timeout(Delay::new(self.config.contract_call_timeout)),
+					0,
+				),
+			},
+		);
+
+		self.waker.wake();
```

Comment:
Can you explain the need in comments to implement our own wakers/runtime?

Reply:
Could you clarify what you mean by implementing our own Waker or runtime?

Reply:
You have implemented what looks something like coroutines--with wakers and polls--and with very similar semantics to how you would implement async runtimes in rust.